Add display/displayMode toggle on test page

edemaine opened the pull request:
* Fix #1035
* Refactor query parsing code to promote re-use
* Also add support for (unescaped) single-character macros

ylemkimon commented on static/main.js:
```diff
@@ -5,6 +5,14 @@
+
+function argument(keyRegex) {
+    return argumentRegex(keyRegex).exec(window.location.search);
+}
```
Would it be possible to document functions?

kevinbarabash replied:
I would probably refactor things to parse all the URL params to an object using a library from npm.  Then we could have simpler regexes for just the keys to check if there are any macros and the rest become simple key lookups.

edemaine replied:
This has turned out to simplify the code a lot.  Good idea, @kevinbarabash!

ylemkimon commented on static/main.js:
```diff
@@ -13,32 +21,38 @@ function init() {
 
+    // Use `display=0` or `displayMode=0` to turn off displayMode.
+    if ((match = argument("display|displayMode"))) {
+        if (match[2] === "false" || match[2] === "0" || match[2] === "no") {
```
I think `n` and maybe `f` are more common.

kevinbarabash replied:
I think these are fine.  TBH I would've been okay with only checking for `displayMode=0`.

edemaine replied:
I added support for anything starting with 0, n, or f.  Hopefully less guesswork in the future.

ylemkimon commented on static/main.js:
```diff
@@ -13,32 +21,38 @@ function init() {
     const options = {displayMode: true};
-    const macroRegex = /(?:^\?|&)(?:\\|%5[Cc])([A-Za-z]+)=([^&]*)/g;
+    const macroRegex = argumentRegex("(?:\\\\|%5[Cc])[A-Za-z]+|.");
     let macroString = "";
```
The name `macroString` seems to be misleading.

kevinbarabash replied:
Maybe `optionsString` may be more appropriate.

edemaine replied:
I've rewritten the code to avoid this recopying of the search string.

kevinbarabash commented on static/main.js:
```diff
@@ -13,32 +21,38 @@ function init() {
-    // https://github.com/Khan/KaTeX/issues/1035
     const options = {displayMode: true};
-    const macroRegex = /(?:^\?|&)(?:\\|%5[Cc])([A-Za-z]+)=([^&]*)/g;
+    const macroRegex = argumentRegex("(?:\\\\|%5[Cc])[A-Za-z]+|.");
```
I didn't realize that we supported macros as URL params, that's pretty cool.

edemaine commented:
Good comments. Most of these are inherited from the old code, but this is probably a fine time to clean it up.

edemaine commented:
I implemented the comments, which cleaned up the code a lot. I also added some small features:
* `pre` and `post are aliases for `before` and `after`. This matches the option names in the screenshotter, so seems nice.
* Macros with escaped characters should now work, thanks to the `query-string` package.

I have not implemented a GUI for these options yet; maybe best for another PR?

kevinbarabash commented:
> I have not implemented a GUI for these options yet; maybe best for another PR?

SGTM

ylemkimon commented:
The file permissions of `package.json` and `package-lock.json` seems to be changed.

edemaine commented:
@ylemkimon Thanks for spotting that! (darn Windows)  Should be fixed now.

kevinbarabash commented:
@ylemkimon thanks for reviewing.  Feel free to merge PRs.
